include/EVT/dev/encoder.hpp

```diff
@@ -0,0 +1,112 @@
+#ifndef _EVT_ENCODER_
+#define _EVT_ENCODER_
+
+#include <EVT/io/GPIO.hpp>
+
+namespace EVT::core::DEV {
+class Encoder {
+public:
+    /**
+     * Constructs an Encoder instance. Call update() within the main loop to update encoder values.
+     * @param a a pin of the encoder
+     * @param b b pin of the encoder
+     * @param range range of the encoder positions (should be positive)
+     * @param initialPosition initial position the encoder is in
+     */
+    Encoder(IO::GPIO& a, IO::GPIO& b, int64_t range, int64_t initialPosition);
+
+    /**
+     * An Enum that represents the direction the encoder is being spun in.
+     * The direction should reset to Static after a couple of reads with no changes.
+     */
+    enum Direction {
+        Left = -1,
+        Static = 0,
+        Right = 1
+    };
+
+    /**
+     * Reads and updates the encoder rotation and the absolute position.
+     * @return what position was changed by, should be in this range: [-3,3]
+     */
+    int8_t update();
+
+    /**
+     * returns the current absolute position
+     * @return the current position of the encoder, between -range and range, inclusive
+     */
+    int64_t getPosition();
+
+    /**
+     * returns the noChangeCounter value;
+     * @return the noChangeCounter value
+     */
+    uint8_t getNoChangeCounter();
+
+    int8_t getRelativePosition();
+
+private:
+    ///GPIO pin a
+    IO::GPIO& a;
+    ///GPIO pin b
+    IO::GPIO& b;
+
+    ///position can be between = [-range, range]
+    int64_t range;
```

Comment:
I'm beginning to rethink this whole range idea. What if people want a range from -10 to 20? What if they want to increment by 2 each time? What if they want an exponential scale? There's no way to support everything someone might want to do, so I think it might be better to support the minimum and then let the user of the class deal with the details.

Here's a possible implementation: the constructor takes in a max value, an initial position, and a "rollover" boolean. The encoder position then only goes between 0 and the max value, so it can be unsigned. Initial position works as currently implemented. The user of the class can then map encoder values to their values using any implementation they want, so we don't have to support it. The only other parameter I think we might need is rollover, which says if we can roll over from the max value back to 0, which could be useful for selecting from a list of options.

Does this implementation make sense? Are there any features you would add? Do you think it would be better to keep it the way it is?

Reply:
Implemented it that way instead, it does make more sense.